### ISSUE TITLE
Install tool with `display_erros` enabled

### ISSUE THREAD

**Issue by richardhj** [contao/core#8666]
The install tool (`contao/install.php`) enables the `display_errors` ini value.
On a live website with `displayErrors` disabled in the `localconfig`, the database was down. While Contao displayed the `be_error` template as expected, I wanted to take a look at the install tool, which displayed the error stack trace like
```
Warning: mysqli::real_connect(): (HY000/2002): Connection refused in system/modules/core/library/Contao/Database/Mysqli.php on line 46
#0 [internal function]: __error(2, 'mysqli::real_co...', '/var/www/virtua...', 46, Array)
#1 system/modules/core/library/Contao/Database/Mysqli.php(46): mysqli->real_connect('localhost', 'username', 'password', 'database', 3306, '')
#2 system/modules/core/library/Contao/Database.php(81): Contao\Database\Mysqli->connect()
#3 system/modules/core/library/Contao/Database.php(165): Contao\Database->__construct(Array)
#4 [internal function]: Contao\Database::getInstance()
#5 system/modules/core/library/Contao/System.php(140): call_user_func(Array)
#6 system/modules/core/classes/Backend.php(30): Contao\System->import('Database')
#7 composer/vendor/menatwork/piwiktrackingtag/system/modules/PiwikTrackingTag/PiwikTrackingTag.php(32): Contao\Backend->__construct()
#8 system/modules/core/library/Contao/System.php(140): PiwikTrackingTag->__construct()
#9 system/modules/core/classes/BackendTemplate.php(41): Contao\System->import('PiwikTrackingTa...')
#10 system/modules/core/classes/BackendTemplate.php(107): Contao\BackendTemplate->parse()
#11 system/modules/core/controllers/BackendInstall.php(878): Contao\BackendTemplate->output()
#12 system/modules/core/controllers/BackendInstall.php(107): Contao\BackendInstall->outputAndExit()
#13 contao/install.php(24): Contao\BackendInstall->run()
#14 {main}

Fatal error: Uncaught exception Exception with message Connection refused thrown in system/modules/core/library/Contao/Database/Mysqli.php on line 50
#0 system/modules/core/library/Contao/Database.php(81): Contao\Database\Mysqli->connect()
#1 system/modules/core/library/Contao/Database.php(165): Contao\Database->__construct(Array)
#2 [internal function]: Contao\Database::getInstance()
#3 system/modules/core/library/Contao/System.php(140): call_user_func(Array)
#4 system/modules/core/classes/Backend.php(30): Contao\System->import('Database')
#5 composer/vendor/menatwork/piwiktrackingtag/system/modules/PiwikTrackingTag/PiwikTrackingTag.php(32): Contao\Backend->__construct()
#6 system/modules/core/library/Contao/System.php(140): PiwikTrackingTag->__construct()
#7 system/modules/core/classes/BackendTemplate.php(41): Contao\System->import('PiwikTrackingTa...')
#8 system/modules/core/classes/BackendTemplate.php(107): Contao\BackendTemplate->parse()
#9 system/modules/core/controllers/BackendInstall.php(878): Contao\BackendTemplate->output()
#10 system/modules/core/controllers/BackendInstall.php(107): Contao\BackendInstall->outputAndExit()
#11 contao/install.php(24): Contao\BackendInstall->run()
#12 {main}
```

As you might see, the database credentials are shown in plain text in the install tool, which is not protected with a password at this point.
The third party app "PiwikTrackingTag" does not influence this behavior.

**Comment by leofeyer** [comment]
> which is not protected with a password at this point

This is not correct. The install tool will not try to enable a database connection if there is no user logged in. I just tested it and the error message is only shown **after** I logged in.

**Comment by fritzmg** [comment]
I can reproduce the issue. The error message
```
Warning: mysqli::real_connect(): (HY000/1045): Access denied for user '…'@'…' (using password: YES) in system\modules\core\library\Contao\Database\Mysqli.php on line 46
```
is shown immediately when accessing the Install Tool.

**Comment by fritzmg** [comment]
I _can_ reproduce this in a non-vanilla Contao 3.5.24 installation __with__ `coreOnlyMode`.
I __cannot__ reproduce this in a _vanilla_ Contao 3.5.24 installation.

**Comment by discordier** [comment]
@fritzmg Can you get a stack trace for the connection attempt from both versions and compare them?

**Comment by fritzmg** [comment]
In one case, the culprit is the `selectModule` extension:
```
#0 [internal function]: __error(2, 'mysqli::real_co...', 'C:\\xampp\\htdocs...', 46, Array)
#1 system\modules\core\library\Contao\Database\Mysqli.php(46): mysqli->real_connect('[…]', '[…]', '[…]', '[…]', 3306, '')
#2 system\modules\core\library\Contao\Database.php(81): Contao\Database\Mysqli->connect()
#3 system\modules\core\library\Contao\Database.php(165): Contao\Database->__construct(Array)
#4 [internal function]: Contao\Database::getInstance()
#5 system\modules\core\library\Contao\System.php(140): call_user_func(Array)
#6 system\modules\core\classes\Backend.php(30): Contao\System->import('Database')
#7 system\modules\core\library\Contao\System.php(140): Contao\Backend->__construct()
#8 system\modules\core\classes\BackendTemplate.php(41): Contao\System->import('SelectModuleHel...')
#9 system\modules\core\classes\BackendTemplate.php(107): Contao\BackendTemplate->parse()
#10 system\modules\core\controllers\BackendInstall.php(878): Contao\BackendTemplate->output()
#11 system\modules\core\controllers\BackendInstall.php(107): Contao\BackendInstall->outputAndExit()
#12 contao\install.php(24): Contao\BackendInstall->run()
#13 {main}
```
After disabling `selectModule`, the culprit is `syncCto`:
```
#0 [internal function]: __error(2, 'mysqli::real_co...', 'C:\\xampp\\htdocs...', 46, Array)
#1 system\modules\core\library\Contao\Database\Mysqli.php(46): mysqli->real_connect('[…]', '[…]', '[…]', '[…]', 3306, '')
#2 system\modules\core\library\Contao\Database.php(81): Contao\Database\Mysqli->connect()
#3 system\modules\core\library\Contao\Database.php(165): Contao\Database->__construct(Array)
#4 [internal function]: Contao\Database::getInstance()
#5 system\modules\core\library\Contao\System.php(140): call_user_func(Array)
#6 system\modules\core\library\Contao\User.php(158): Contao\System->import('Database')
#7 system\modules\core\classes\BackendUser.php(102): Contao\User->__construct()
#8 system\modules\core\library\Contao\User.php(220): Contao\BackendUser->__construct()
#9 composer\vendor\menatwork\synccto\system\modules\syncCto\SyncCtoHelper.php(61): Contao\User::getInstance()
#10 composer\vendor\menatwork\synccto\system\modules\syncCto\SyncCtoHelper.php(101): SyncCtoHelper->__construct()
#11 [internal function]: SyncCtoHelper::getInstance()
#12 system\modules\core\library\Contao\System.php(140): call_user_func(Array)
#13 system\modules\core\classes\BackendTemplate.php(41): Contao\System->import('SyncCtoHelper')
#14 system\modules\core\classes\BackendTemplate.php(107): Contao\BackendTemplate->parse()
#15 system\modules\core\controllers\BackendInstall.php(878): Contao\BackendTemplate->output()
#16 system\modules\core\controllers\BackendInstall.php(107): Contao\BackendInstall->outputAndExit()
#17 contao\install.php(24): Contao\BackendInstall->run()
#18 {main}
```
After disabling `syncCto`, no error will be displayed and you can log into the Install Tool normally. Also I think I was wrong about the `coreOnlyMode` - when I enabled it now, I was able to access the Install Tool normally, may be I had a typo before.

Basically any extension that implements a `parseBackendTemplate` Hook can cause this problem - at least if the constructor of the class of the Hook initializes the database instance in one way or another.

In case of `selectModule` it happens because the class of the hook extends `Contao\Backend`. In `syncCto` it happens because the constructor of the class of the hook initializes the `Contao\BackendUser` instance.

So the question is: should extensions be careful to _not_ initialize any database instance directly or indirectly in their `parseBackendTemplate`? Or should this be something that the Core or the Install Tool should handle in a better way.

The former is not feasible imho.

**Comment by fritzmg** [comment]
> The third party app "PiwikTrackingTag" does not influence this behavior.

That is false ;). It's the same story here. "PiwikTrackingTag" uses a `parseBackendTemplate` Hook, which in turn extends `Contao\Backend` which in turn initializes the database instance.

**Comment by fritzmg** [comment]
May be the `parseBackendTemplate` Hook should not be executed for the Install Tool and the Backend Login in general. There was a (minor) security issue with `menatwork/MultiColumnWizard` wizard, where through the `parseBackendTemplate` Hook the exact version of the Contao installation was visible to the public through the Install Tool and the Backend Login, see https://github.com/menatwork/MultiColumnWizard/issues/184

This was subsequently fixed in https://github.com/menatwork/MultiColumnWizard/pull/213 by checking the template name for `be_login` and `be_install`, so that this version information is not added in the Backend Login or the Install Tool. But may be it is a better idea in general to not allow the alteration of the Backend Login or Install Tool template through a Hook.

**Comment by richardhj** [comment]
> >The third party app "PiwikTrackingTag" does not influence this behavior.

> That is false ;)

Hehe, I just disabled *PiwikTrackingTag* and noticed the stack trace corresponding to *syncCto*, just as you experienced it. At this point I did not dig into the code. So nice analysis @fritzmg

**Comment by richardhj** [comment]
# Proposed solutions
## Disable hooks
```diff
	/**
	 * Output the template file and exit
	 */
	protected function outputAndExit()
	{
+	    // Do not load the hooks as they might expose the database credentials. See #8666
+		if (!$this->checkIsAuthenticated())
+		{
+			unset($GLOBALS['TL_HOOKS']['parseBackendTemplate']);
+			unset($GLOBALS['TL_HOOKS']['outputBackendTemplate']);
+		}
+
		$this->Template->theme = \Backend::getTheme();
		$this->Template->base = \Environment::get('base');
		$this->Template->language = $GLOBALS['TL_LANGUAGE'];
		$this->Template->charset = \Config::get('characterSet');
		$this->Template->pageOffset = \Input::cookie('BE_PAGE_OFFSET');
		$this->Template->action = ampersand(\Environment::get('request'));
		$this->Template->noCookies = $GLOBALS['TL_LANG']['MSC']['noCookies'];
		$this->Template->title = specialchars($GLOBALS['TL_LANG']['tl_install']['installTool'][0]);
		$this->Template->expandNode = $GLOBALS['TL_LANG']['MSC']['expandNode'];
		$this->Template->collapseNode = $GLOBALS['TL_LANG']['MSC']['collapseNode'];
		$this->Template->loadingData = $GLOBALS['TL_LANG']['MSC']['loadingData'];
		$this->Template->ie6warning = sprintf($GLOBALS['TL_LANG']['ERR']['ie6warning'], '<a href="http://ie6countdown.com">', '</a>');
		$this->Template->hasComposer = is_dir(TL_ROOT . '/system/modules/!composer');

		$this->Template->output();
		exit;
	}
```
## Change the way enabling display_errors
or simply set the value if a user is authenticated
```diff
+		if ($this->checkIsAuthenticated())
+		{
+			// Show error messages
			@ini_set('display_errors', 1);
			error_reporting(Config::get('errorReporting'));
+		}
```

```php
    /**
     * Check whether the install tool is accessed authenticated
     *
     * @return bool
     */
	protected function checkIsAuthenticated()
	{
		return (\Input::cookie('TL_INSTALL_AUTH') && $_SESSION['TL_INSTALL_AUTH'] != '' && \Input::cookie('TL_INSTALL_AUTH') == $_SESSION['TL_INSTALL_AUTH'] && !$_SESSION['TL_INSTALL_EXPIRE'] < time());
	}
```

**Comment by leofeyer** [comment]
Fixed in 5cb339dc04a4e34478e168834f930bb07ff45b2e.

**Comment by leofeyer** [comment]
Thank you @fritzmg and @richardhj.
